text/00110-cli-framework-compatibility-strategy.md

```diff
@@ -0,0 +1,464 @@
+
+This creates clutter, and might also cause issues with applying `jsii` compatibility checks.
+
+In addition, according to our new compatibility model, we need to make sure the user gets the correct error
+when the CLI version is smaller than the `Cloud-Assembly` version.
+
+The plan therefore is as follows:
+
+### Step 1: Create a separate package called `cx-protocol`
+
+This package will provide:
+
+#### (1) Object Model
+
+- CloudAssembly
+- RuntimeInfo
+- AssemblyManifest
+- CloudArtifact
+- ...
+- ...
+
+These are the objects that will be later used by the CLI. Running compatibility checks on them **should** ensure:
+
+1. Rename/Remove properties is not allowed.
+2. Changing the type of a property is not allowed.
+3. Making an optional property required is not allowed.
+4. Adding a required property is not allowed.
+
+> The `jsii` compatibility checker currently doesn't actually provide us with all those requirements. It validates interfaces only with relation to their
+> usage, in either argument or return values of methods. This means that, for example, number 4 is allowed,
+> because the `ContainerImageAssetMetadataEntry` is not used as a function argument in `cx-api`.
+> In order to achieve this, we will either need to enhance the `jsii` checker, or refactor our code.
+In any case, we stipulate the necessary changes will be done.
+
+##### (2) De-serialization
+
+The CLI will need to create this object model in-memory, from the disk representation.
+For this, we will provide a de-serialization method:
+
+```typescript
+const assembly = new CloudAssembly("/path/to/cloud-assembly");
+const artifacts = assembly.artifacts;
+const version = assembly.version;
+const runtime = assembly.runtime;
+```
+
+This method will also be used by the framework, as the return value of the `app.synth()` method.
+
+> Serialization will not be part of this package, as it is only used in the framework.
+
+This new package has to be **stable** to enforce `jsii` compatibility. This is ok, it actually makes perfect
+sense for it to be stable as it represents the contract the `Cloud-Assembly` needs to adhere to.
+
+#### Step 2: Change CLI dependencies
+
+The CLI should only depend on the new `cx-protocol` package, and **not** on the `cx-api` package.
+
+This means that when users install new versions of the CLI, they also get a new version of `cx-protocol`,
+but since `cx-protocol` is backwards compatible, there is no risk of breaking API.
+
+> The CLI currently depends on parts of `cx-api` that don't strictly relate to the proposed `cx-protocol` object model.
+> Analysis of those dependencies has yet to be done. Ideally, we would be able to move that code away from `cx-api` and into the CLI itself. Perhaps even at the expense of some duplication if that code is being used in the framework as well.
+> In any case, I don't think this will be a blocker for the design.
```

Comment:
So basically the proposal is to get rid of cx-api altogether? 

What about all the other stuff in there?

Reply:
I would very much like to get rid of `cx-api`. But this RFC doesn't strictly propose that.  

I'm still in the process of understanding where to put all the other stuff. I haven't yet mapped all of this out. If I am able to get rid of it rather easily, i will do it in the PR implementing this RFC, if not, its ok if we leave this for now as a shared code graveyard, and slowly pick it apart. 

So basically, nice to have.